### ISSUE TITLE
BUG: Docketing a file from a message inserts the message Received date as the Date on the docket record

### ISSUE THREAD

**Issue by cholly75** [ustaxcourt/ef-cms#8373]
**Describe the Bug**
When a docket clerk serves a court issued document attached to a message, the resulting docket entry displays the incorrect value in the `Date` column of the Docket Record. 

Prior to 12/5 the Date field displayed the served date as expected.

Example: Docket 12146-20, ustaxcourt/ef-cms#6507

**Business Impact/Reason for Severity**
High - clerks must edit the Filed date on the docket entry to correct the record, and most docket work is done via messaging.

**In which environment did you see this bug?**
PROD, TEST

**Who were you logged in as?**
docketclerk4

**What were you doing when you discovered this bug? (Using the application, demoing, smoke tests, testing other functionality, etc.)**
Checking production report

**To Reproduce**
1. Go to 'Section Messages' as a docket clerk
2. Find a message containing a court issued document that has NOT yet been served. (To make testing easier, verify the date the document was filed is not also today).
3. Open the message and add the document to the record, saving and serving it.
4. Check the docket record
5. Observe the `Date` column does NOT match the `Served` column for the newly added entry in the Docket Record

**Expected Behavior**
`Date` field on the docket record for a court issued document document has same value as the `Served` column.

**Actual Behavior**
Date field on the docket record for the court issued document appears to display the date the document was filed.

**Screenshots**

**Desktop (please complete the following information):**

**Smartphone (please complete the following information):**

**Cause of Bug, If Known**
This bug was caused by removing explicitly setting `filingDate` to "now" in `fileAndServeCourtIssuedInteractor`. The reason that value was removed was because it appeared at the time to be duplicative logic as `filingDate` for court issued documents should be set to the same time the document is served, and the DocketEntry constructor has a default to cover that case. 

Zooming out a bit, we need to investigate the business rules around filingDate and servedDate for court issued documents. If there is a business need for filingDate to always be servedDate for court issued documents, then we need to make updates to the relevant entities. If this is a presentation issue, we can modify the presenter for docket record to reflect this need.

**Process for Logging a Bug:**
* Complete the above information
* Add a severity tag (Critical, High Severity, Medium Severity or Low Severity). See below for priority definition. 

**Severity Definition:**
* Critical Defect
Blocks entire system's or module’s functionality
No workarounds available
Testing cannot proceed further without bug being fixed.

* High-severity Defect
Affects key functionality of an application
There's a workaround, but not obvious or easy
App behaves in a way that is strongly different from the one stated in the requirements

* Medium-severity Defect
A minor function does not behave in a way stated in the requirements.
Workaround is available and easy

* Low-severity Defect
Mostly related to an application’s UI
Doesn't need a workaround, because it doesn't impact functionality

## Definition of Ready for Bugs(Created 10-4-21)
Definition used: A failure or flaw in the system which produces an incorrect or undesired result that deviates from the expected result or behavior. (Note: Expected results are use cases that have been documented in past user stories as acceptance criteria and test cases, and do not include strange behavior unrelated to use cases.)

The following criteria must be met in order for the development team to begin work on the bug.

The bug must:
 - Be focused on solving a user problem
 - Contain data for all fields in the bug template, so the team can pick it up and begin working immediately

Process: If the unexpected results are new use cases that have been identified, but not yet built, new acceptance criteria and test cases should be captured in a new user story and prioritized by the product owner. 

If the Court is not able to reproduce the bug, add the “Unable to reproduce” tag. This will provide visibility into the type of support that may be needed by the Court. In the event that the Court cannot reproduce the bug, the Court will work with Flexion to communicate what type of troubleshooting help may be needed.

## Definition of Done (Updated 4-14-21)
**Product Owner**
 - [ ]  Bug fix has been validated in the Court's test environment

**Engineering**
 - [x] Automated test scripts have been written
 - [x] Field level and page level validation errors (front-end and server-side) integrated and functioning
 - [x] Verify that language for docket record for internal users and external users is identical
 - [x] New screens have been added to pa11y scripts
 - [x] All new functionality verified to work with keyboard and macOS voiceover https://www.apple.com/voiceover/info/guide/_1124.html 
 - [x] READMEs, other appropriate docs, JSDocs and swagger/APIs fully updated
 - [x] UI should be touch optimized and responsive for external only (functions on supported mobile devices and optimized for screen sizes as required)
 - [x] Interactors should validate entities before calling persistence methods
 - [x] Code refactored for clarity and to remove any known technical debt
 - [x] Deployed to the Court's `test` environment if prod-like data is required. Otherwise, deployed to any `experimental` environment for review.

**Comment by rachelschneiderman** [comment]
After speaking with @cholly75 we clarified some business rules.

Unservable court issued documents `filingDate` is the day the document is added to the docket record.  Servable court issued documents `filingDate` should be the same as the `servedDate`.

For now, we will revert the change that caused the bug and merge that ASAP. For a more maintainable, long term solution we will update entity validation to match the rules above as a separate task for this bug.

**Comment by codyseibert** [comment]
We had a discussion with @cholly75 in regards to refactoring some business logic / entity validation rules to verify the filingDate matches the servedAt timestamp.  The issue we run into is that dawson internal users currently have the ability to manually change the filing date (we are not sure why they have that feature for court issued documents), so Holly plans to speak with some employees of the court to find out if manually changing the filingDate is necessary and figure out any edge case involved with this change.

**Comment by codyseibert** [comment]
blocking until more information is provided around the business rules about court issued documents. The original temp fix can still be deployed, but we want to put this story back in progress if we determine more refactoring is needed

**Comment by rachelschneiderman** [comment]
Moving back to in-progress, still awaiting clarification on business need for filing date/served date rules for court issued documents.

**Comment by rachelschneiderman** [comment]
RE: editing `filingDate` for court issued documents. The court has determined there is a need to be able to edit `filingDate` for minute entries (ex - Filing Fee Paid) ONLY, all other court issued documents should have `filingDate` equal to `servedAt` and the ability to edit `filingDate` could be removed.

Before we can make a decision on whether or not to enforce that as a rule, the court needs to know how many docket entries that change would affect as some docket records have been entered as evidence in cases and should not be changed.

**Comment by rachelschneiderman** [comment]
A script was written to determine how many court issued docket entries had `servedAt` dates that differed from `filingDate` and run against a recently glued environment. There are thousands of docket entries that fit into that category, therefore, we can't enforce stricter validation for `filingDate`. Instead, we proposed the following work as a way to get around needing to migrate data: 

- Update which value is displayed in the Docket Record table for served court issued docket entries - instead of `filingDate`, display `servedAt`
- Update validation rules for DocketEntry to allow `filingDate` to be undefined for court issued docket entries
- Remove the ability to edit filing date for court issued docket entries
- When serving a court issued docket entry, set the `filingDate` to undefined instead of "now"

We will move these proposed tasks to a new user story as there is concern about the implications of changing how we set properties on docket entries in production without a migration.
